File: src/core/KeyManager.ts
export interface KeyAsset {
  label: string;
  account: string;
  key: string;
}

class KeyPool {
  private deck: KeyAsset[] = [];
  private pointer: number = 0;
  private type: string;

  constructor(envString: string | undefined, type: string) {
    this.type = type;
    if (!envString) {
      console.warn(`\x1b[33m[⚠️] NO KEYS LOADED FOR ${type}\x1b[0m`);
      return;
    }

    const entries = envString.split(',');
    entries.forEach((entry, idx) => {
      let label = "";
      let key = "";
      
      if (entry.includes(':')) {
        const parts = entry.split(':');
        label = parts[0];
        key = parts[1];
      } else {
        label = `${type}_DEALER_${String(idx + 1).padStart(2, '0')}`;
        key = entry;
      }
      
      this.deck.push({ 
        label: label.trim(), 
        account: label.trim(), 
        key: key.trim() 
      });
    });
    this.shuffle();
  }

  private shuffle() {
    if (this.deck.length === 0) return;
    console.log(`\n\x1b[1;33m[🎲] ${this.type} DECK SHUFFLING...\x1b[0m`);
    for (let i = this.deck.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [this.deck[i], this.deck[j]] = [this.deck[j], this.deck[i]];
    }
    this.pointer = 0;
  }

  public getNext(): KeyAsset {
    if (this.deck.length === 0) throw new Error(`NO AMMUNITION FOR ${this.type}`);
    const asset = this.deck[this.pointer];
    this.pointer++;
    if (this.pointer >= this.deck.length) this.shuffle();
    return asset;
  }

  public dump() {
    console.log(`\n\x1b[1;35m--- [ ${this.type} ARSENAL LOADED ] ---\x1b[0m`);
    this.deck.forEach((a, i) => {
        const masked = a.key.length > 8 ? `${a.key.substring(0, 8)}...` : 'INVALID';
        console.log(`[${String(i+1).padStart(2,'0')}] \x1b[1;92m${a.account.padEnd(20)}\x1b[0m | ID: ${masked}`);
    });
  }
}

// Load pools from Environment Variables
export const GroqPool = new KeyPool(process.env.GROQ_KEYS, 'GROQ');
export const GooglePool = new KeyPool(process.env.GOOGLE_KEYS, 'GOOGLE');
export const DeepSeekPool = new KeyPool(process.env.DEEPSEEK_KEYS, 'DEEPSEEK');
export const MistralPool = new KeyPool(process.env.MISTRAL_KEYS, 'MISTRAL');

GroqPool.dump();
GooglePool.dump();
DeepSeekPool.dump();
MistralPool.dump();

File: app/core/key_manager.py
"""
PEACOCK ENGINE - Key Manager
Handles API key rotation, shuffling, and usage tracking.
"""

import os
import random
from typing import List, Optional, Tuple
from dotenv import load_dotenv
from pydantic import BaseModel

# Ensure environment is loaded
load_dotenv()

from app.utils.formatter import CLIFormatter, Colors
from app.db.database import KeyUsageDB


import time

class KeyAsset(BaseModel):
    label: str
    account: str
    key: str
    cooldown_until: float = 0.0

    @property
    def on_cooldown(self) -> bool:
        return time.time() < self.cooldown_until


from abc import ABC, abstractmethod

class RotationStrategy(ABC):
    @abstractmethod
    def get_next(self, deck: List[KeyAsset], pointer: int) -> Tuple[KeyAsset, int]:
        pass

class ShuffleStrategy(RotationStrategy):
    """Original 'Deck of Cards' logic."""
    def get_next(self, deck: List[KeyAsset], pointer: int) -> Tuple[KeyAsset, int]:
        asset = deck[pointer]
        new_pointer = pointer + 1
        return asset, new_pointer

class RoundRobinStrategy(RotationStrategy):
    """Simple 1, 2, 3 rotation."""
    def get_next(self, deck: List[KeyAsset], pointer: int) -> Tuple[KeyAsset, int]:
        asset = deck[pointer]
        new_pointer = (pointer + 1) % len(deck)
        return asset, new_pointer

class KeyPool:
    def __init__(self, env_string: Optional[str], pool_type: str):
        self.deck: List[KeyAsset] = []
        self.pointer: int = 0
        self.pool_type: str = pool_type
        self.strategy: RotationStrategy = ShuffleStrategy()
        
        if not env_string:
            CLIFormatter.warning(f"NO KEYS LOADED FOR {pool_type}")
            return

        entries = env_string.split(',')
        for idx, entry in enumerate(entries):
            label = ""
            key = ""
            
            if ':' in entry:
                parts = entry.split(':')
                label = parts[0]
                key = parts[1]
            else:
                label = f"{pool_type}_DEALER_{str(idx + 1).zfill(2)}"
                key = entry
            
            self.deck.append(KeyAsset(
                label=label.strip(),
                account=label.strip(),
                key=key.strip()
            ))
        self.shuffle()
        CLIFormatter.success(f"{pool_type} POOL: {len(self.deck)} KEYS LOADED")

    def set_strategy(self, strategy_name: str):
        if strategy_name == "round_robin":
            self.strategy = RoundRobinStrategy()
        else:
            self.strategy = ShuffleStrategy()

    def shuffle(self):
        if not self.deck:
            return
        if os.getenv("PEACOCK_VERBOSE") == "true":
            style = CLIFormatter.get_gateway_style(self.pool_type)
            print(f"{style['color']}[🎲] {self.pool_type} DECK SHUFFLING...{Colors.RESET}")
        random.shuffle(self.deck)
        self.pointer = 0

    def mark_cooldown(self, account: str, duration: int = 60):
        """Mark a specific key as being on cooldown."""
        for asset in self.deck:
            if asset.account == account:
                asset.cooldown_until = time.time() + duration
                CLIFormatter.warning(f"KEY [{account}] marked on COOLDOWN for {duration}s")
                return

    def get_next(self) -> KeyAsset:
        if not self.deck:
            raise Exception(f"NO AMMUNITION FOR {self.pool_type}")
        
        # Try to find a key not on cooldown (up to a full deck rotation)
        for _ in range(len(self.deck)):
            asset, self.pointer = self.strategy.get_next(self.deck, self.pointer)
            
            # If we hit the end of a shuffle deck, reshuffle
            if isinstance(self.strategy, ShuffleStrategy) and self.pointer >= len(self.deck):
                self.shuffle()
            
            if not asset.on_cooldown:
                return asset
                
        # All keys are on cooldown!
        raise Exception(f"ALL KEYS ON COOLDOWN FOR {self.pool_type.upper()}")

    def dump(self):
        style = CLIFormatter.get_gateway_style(self.pool_type)
        print(f"\n{style['color']}--- [ {self.pool_type} ARSENAL LOADED ] ---{Colors.RESET}")
        for i, a in enumerate(self.deck):
            masked = f"{a.key[:8]}..." if len(a.key) > 8 else "INVALID"
            print(f"{style['color']}[{str(i+1).zfill(2)}]{Colors.RESET} {Colors.YELLOW}{a.account.ljust(20)}{Colors.RESET} | ID: {masked}")

    @staticmethod
    def record_usage(gateway: str, account: str, usage: dict):
        """Record usage for a key to the database."""
        try:
            KeyUsageDB.record_usage(gateway, account, usage)
        except Exception as e:
            CLIFormatter.warning(f"Failed to record key usage: {e}")


# Load pools from Environment Variables
GroqPool = KeyPool(os.getenv("GROQ_KEYS"), "groq")
GooglePool = KeyPool(os.getenv("GOOGLE_KEYS"), "google")
DeepSeekPool = KeyPool(os.getenv("DEEPSEEK_KEYS"), "deepseek")
MistralPool = KeyPool(os.getenv("MISTRAL_KEYS"), "mistral")

if os.getenv("PEACOCK_DEBUG") == "true":
    GroqPool.dump()
    GooglePool.dump()
    DeepSeekPool.dump()
    MistralPool.dump()

